File: LeetCode/DP1D/python/lengthOfLIS.py
from dataclasses import dataclass
from enum import Enum
from typing import List, Tuple


class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"


@dataclass
class LengthOfLISRecord:
    nums: List[int]
    expected: int


class Solution:
    def computeLengthOfLIS(self, nums: List[int]) -> int:
        LIS = [1] * len(nums)

        for i in range(len(nums) - 1, -1, -1):
            for j in range(i + 1, len(nums)):
                if nums[i] < nums[j]:
                    LIS[i] = max(LIS[i], 1 + LIS[j])

        return max(LIS)


def testSolution(record: LengthOfLISRecord) -> None:
    print(f"nums: {record.nums}")
    print(f"expected: {record.expected}")

    result = Solution().computeLengthOfLIS(record.nums)

    print(f"result: {result}")
    print(Result.PASS.value if result == record.expected else Result.FAIL.value)


if __name__ == "__main__":
    records: Tuple[LengthOfLISRecord, ...] = (
        LengthOfLISRecord([10, 9, 2, 5, 3, 7, 101, 18], 4),
        LengthOfLISRecord([0, 1, 0, 3, 2, 3], 4),
        LengthOfLISRecord([7, 7, 7, 7, 7, 7, 7], 1),
    )

    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        testSolution(record)
        print("-" * 50)

File: LeetCode/DP1D/JavaScript/lengthOfLIS.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

class LengthOfLISRecord {
  constructor(nums, expected) {
    this.nums = nums;
    this.expected = expected;
  }
}

class Solution {
  /**
   * @param {number[]} nums
   * @return {number}
   */
  lengthOfLIS(nums) {
    const lis = Array(nums.length).fill(1);

    for (let i = nums.length - 1; i > -1; i -= 1) {
      for (let j = i + 1; j < nums.length; j += 1) {
        if (nums[i] < nums[j]) {
          lis[i] = Math.max(lis[i], 1 + lis[j]);
        }
      }
    }

    return Math.max(...lis);
  }
}

function testSolution(record) {
  const solution = new Solution();
  const result = solution.lengthOfLIS(record.nums);
  const status = result === record.expected ? Result.PASS : Result.FAIL;

  console.log(`Input: ${JSON.stringify(record.nums)}`);
  console.log(`Expected: ${record.expected}`);
  console.log(`Result: ${result}`);
  console.log(status);
}

const records = [
  new LengthOfLISRecord([10, 9, 2, 5, 3, 7, 101, 18], 4),
  new LengthOfLISRecord([0, 1, 0, 3, 2, 3], 4),
  new LengthOfLISRecord([7, 7, 7, 7, 7, 7, 7], 1),
];

records.forEach((record, index) => {
  console.log(`# Test case ${index + 1}`);
  testSolution(record);
  console.log("-".repeat(50));
});
